File: dqops/src/main/java/com/dqops/checks/table/partitioned/TableMonthlyPartitionedCheckCategoriesSpec.java
/*
 * Copyright © 2021-Present DQOps, Documati sp. z o.o. (support@dqops.com)
 *
 * This file is licensed under the Business Source License 1.1,
 * which can be found in the root directory of this repository.
 *
 * Change Date: This file will be licensed under the Apache License, Version 2.0,
 * four (4) years from its last modification date.
 */
package com.dqops.checks.table.partitioned;

import com.dqops.checks.AbstractRootChecksContainerSpec;
import com.dqops.checks.CheckTarget;
import com.dqops.checks.CheckTimeScale;
import com.dqops.checks.CheckType;
import com.dqops.checks.table.partitioned.comparison.TableComparisonMonthlyPartitionedChecksSpecMap;
import com.dqops.checks.table.partitioned.customsql.TableCustomSqlMonthlyPartitionedChecksSpec;
import com.dqops.checks.table.partitioned.timeliness.TableTimelinessMonthlyPartitionedChecksSpec;
import com.dqops.checks.table.partitioned.uniqueness.TableUniquenessMonthlyPartitionChecksSpec;
import com.dqops.checks.table.partitioned.volume.TableVolumeMonthlyPartitionedChecksSpec;
import com.dqops.metadata.id.ChildHierarchyNodeFieldMap;
import com.dqops.metadata.id.ChildHierarchyNodeFieldMapImpl;
import com.dqops.metadata.scheduling.CheckRunScheduleGroup;
import com.dqops.metadata.sources.TableSpec;
import com.dqops.metadata.timeseries.TimePeriodGradient;
import com.dqops.metadata.timeseries.TimeSeriesConfigurationSpec;
import com.dqops.metadata.timeseries.TimeSeriesMode;
import com.dqops.utils.docs.generators.SampleValueFactory;
import com.dqops.utils.serialization.IgnoreEmptyYamlSerializer;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonPropertyDescription;
import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.fasterxml.jackson.databind.annotation.JsonNaming;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import lombok.EqualsAndHashCode;

import java.util.Objects;

/**
 * Container of table level monthly partitioned checks. Contains categories of monthly partitioned checks.
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
@EqualsAndHashCode(callSuper = true)
public class TableMonthlyPartitionedCheckCategoriesSpec extends AbstractRootChecksContainerSpec {
    public static final ChildHierarchyNodeFieldMapImpl<TableMonthlyPartitionedCheckCategoriesSpec> FIELDS = new ChildHierarchyNodeFieldMapImpl<>(AbstractRootChecksContainerSpec.FIELDS) {
        {
            put("volume", o -> o.volume);
            put("timeliness", o -> o.timeliness);
            put("custom_sql", o -> o.customSql);
            put("uniqueness", o -> o.uniqueness);
            put("comparisons", o -> o.comparisons);

            // accuracy checks are not supported on partitioned checks yet, but we support comparisons
        }
    };

    @JsonPropertyDescription("Volume monthly partitioned data quality checks that verify the quality of every month of data separately")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    @JsonSerialize(using = IgnoreEmptyYamlSerializer.class)
    private TableVolumeMonthlyPartitionedChecksSpec volume;

    @JsonPropertyDescription("Monthly partitioned timeliness checks")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    @JsonSerialize(using = IgnoreEmptyYamlSerializer.class)
    private TableTimelinessMonthlyPartitionedChecksSpec timeliness;

    @JsonPropertyDescription("Custom SQL monthly partitioned data quality checks that verify the quality of every month of data separately")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    @JsonSerialize(using = IgnoreEmptyYamlSerializer.class)
    private TableCustomSqlMonthlyPartitionedChecksSpec customSql;

    @JsonPropertyDescription("Monthly partitioned uniqueness checks on a table level.")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    @JsonSerialize(using = IgnoreEmptyYamlSerializer.class)
    private TableUniquenessMonthlyPartitionChecksSpec uniqueness;

    @JsonPropertyDescription("Dictionary of configuration of checks for table comparisons. The key that identifies each comparison must match the name of a data comparison that is configured on the parent table.")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    @JsonSerialize(using = IgnoreEmptyYamlSerializer.class)
    private TableComparisonMonthlyPartitionedChecksSpecMap comparisons = new TableComparisonMonthlyPartitionedChecksSpecMap();


    /**
     * Returns the container of monthly partitioned checks for volume data quality checks.
     * @return Container of row volume data quality checks.
     */
    public TableVolumeMonthlyPartitionedChecksSpec getVolume() {
        return volume;
    }

    /**
     * Sets the container of volume data quality checks.
     * @param volume New volume checks.
     */
    public void setVolume(TableVolumeMonthlyPartitionedChecksSpec volume) {
        this.setDirtyIf(!Objects.equals(this.volume, volume));
        this.volume = volume;
        this.propagateHierarchyIdToField(volume, "volume");
    }

    /**
     * Returns a container of table level timeliness partitioned checks.
     * @return Custom timeliness partitioned checks.
     */
    public TableTimelinessMonthlyPartitionedChecksSpec getTimeliness() {
        return timeliness;
    }

    /**
     * Sets a reference to a container of timeliness partitioned checks.
     * @param timeliness Custom timeliness partitioned checks.
     */
    public void setTimeliness(TableTimelinessMonthlyPartitionedChecksSpec timeliness) {
        this.setDirtyIf(!Objects.equals(this.timeliness, timeliness));
        this.timeliness = timeliness;
        this.propagateHierarchyIdToField(timeliness, "timeliness");
    }

    /**
     * Returns the container of custom SQL checks.
     * @return Custom sql checks.
     */
    public TableCustomSqlMonthlyPartitionedChecksSpec getCustomSql() {
        return customSql;
    }

    /**
     * Sets a reference to a container of custom SQL checks.
     * @param customSql Container of custom SQL checks.
     */
    public void setCustomSql(TableCustomSqlMonthlyPartitionedChecksSpec customSql) {
        this.setDirtyIf(!Objects.equals(this.customSql, customSql));
        this.customSql = customSql;
        this.propagateHierarchyIdToField(customSql, "custom_sql");
    }

    /**
     * Returns a container of table uniqueness checks.
     * @return Table uniqueness checks.
     */
    public TableUniquenessMonthlyPartitionChecksSpec getUniqueness() {
        return uniqueness;
    }

    /**
     * Sets a reference to a container with the table uniqueness checks.
     * @param uniqueness Container of table uniqueness checks.
     */
    public void setUniqueness(TableUniquenessMonthlyPartitionChecksSpec uniqueness) {
        this.setDirtyIf(!Objects.equals(this.uniqueness, uniqueness));
        this.uniqueness = uniqueness;
        this.propagateHierarchyIdToField(uniqueness, "uniqueness");
    }

    /**
     * Returns the dictionary of comparisons.
     * @return Dictionary of comparisons.
     */
    @Override
    public TableComparisonMonthlyPartitionedChecksSpecMap getComparisons() {
        return comparisons;
    }

    /**
     * Sets the dictionary of comparisons.
     * @param comparisons Dictionary of comparisons.
     */
    public void setComparisons(TableComparisonMonthlyPartitionedChecksSpecMap comparisons) {
        this.setDirtyIf(!Objects.equals(this.comparisons, comparisons));
        this.comparisons = comparisons;
        this.propagateHierarchyIdToField(comparisons, "comparisons");
    }

    /**
     * Returns the child map on the spec class with all fields.
     *
     * @return Return the field map.
     */
    @Override
    protected ChildHierarchyNodeFieldMap getChildMap() {
        return FIELDS;
    }

    /**
     * Returns time series configuration for the given group of checks.
     *
     * @param tableSpec Parent table specification - used to get the details about the time partitioning column.
     * @return Time series configuration.
     */
    @Override
    public TimeSeriesConfigurationSpec getTimeSeriesConfiguration(TableSpec tableSpec) {
        return new TimeSeriesConfigurationSpec()
        {{
            setMode(TimeSeriesMode.timestamp_column);
            setTimeGradient(TimePeriodGradient.month);
            setTimestampColumn(tableSpec.getTimestampColumns().getPartitionByColumn());
        }};
    }

    /**
     * Returns the type of checks (profiling, monitoring, partitioned).
     *
     * @return Check type.
     */
    @Override
    @JsonIgnore
    public CheckType getCheckType() {
        return CheckType.partitioned;
    }

    /**
     * Returns the time range for monitoring and partitioned checks (daily, monthly, etc.).
     * Profiling checks do not have a time range and return null.
     *
     * @return Time range (daily, monthly, ...).
     */
    @Override
    @JsonIgnore
    public CheckTimeScale getCheckTimeScale() {
        return CheckTimeScale.monthly;
    }

    /**
     * Returns the check target, where the check could be applied.
     *
     * @return Check target, "table" or "column".
     */
    @Override
    @JsonIgnore
    public CheckTarget getCheckTarget() {
        return CheckTarget.table;
    }

    /**
     * Returns the name of the cron expression that is used to schedule checks in this check root object.
     *
     * @return Monitoring schedule group (named schedule) that is used to schedule the checks in this root.
     */
    @Override
    @JsonIgnore
    public CheckRunScheduleGroup getSchedulingGroup() {
        return CheckRunScheduleGroup.partitioned_monthly;
    }

    public static class TableMonthlyPartitionedCheckCategoriesSpecSampleFactory implements SampleValueFactory<TableMonthlyPartitionedCheckCategoriesSpec> {
        @Override
        public TableMonthlyPartitionedCheckCategoriesSpec createSample() {
            return new TableMonthlyPartitionedCheckCategoriesSpec() {{
                setVolume(new TableVolumeMonthlyPartitionedChecksSpec.TableVolumeMonthlyPartitionedChecksSpecSampleFactory().createSample());
            }};
        }
    }
}

File: distribution/python/dqops/client/models/table_monthly_partitioned_check_categories_spec.py
from typing import TYPE_CHECKING, Any, Dict, List, Type, TypeVar, Union

from attrs import define as _attrs_define
from attrs import field as _attrs_field

from ..types import UNSET, Unset

if TYPE_CHECKING:
    from ..models.table_custom_sql_monthly_partitioned_checks_spec import (
        TableCustomSqlMonthlyPartitionedChecksSpec,
    )
    from ..models.table_monthly_partitioned_check_categories_spec_comparisons import (
        TableMonthlyPartitionedCheckCategoriesSpecComparisons,
    )
    from ..models.table_monthly_partitioned_check_categories_spec_custom import (
        TableMonthlyPartitionedCheckCategoriesSpecCustom,
    )
    from ..models.table_timeliness_monthly_partitioned_checks_spec import (
        TableTimelinessMonthlyPartitionedChecksSpec,
    )
    from ..models.table_uniqueness_monthly_partition_checks_spec import (
        TableUniquenessMonthlyPartitionChecksSpec,
    )
    from ..models.table_volume_monthly_partitioned_checks_spec import (
        TableVolumeMonthlyPartitionedChecksSpec,
    )


T = TypeVar("T", bound="TableMonthlyPartitionedCheckCategoriesSpec")


@_attrs_define
class TableMonthlyPartitionedCheckCategoriesSpec:
    """
    Attributes:
        custom (Union[Unset, TableMonthlyPartitionedCheckCategoriesSpecCustom]): Dictionary of custom checks. The keys
            are check names within this category.
        volume (Union[Unset, TableVolumeMonthlyPartitionedChecksSpec]):
        timeliness (Union[Unset, TableTimelinessMonthlyPartitionedChecksSpec]):
        custom_sql (Union[Unset, TableCustomSqlMonthlyPartitionedChecksSpec]):
        uniqueness (Union[Unset, TableUniquenessMonthlyPartitionChecksSpec]):
        comparisons (Union[Unset, TableMonthlyPartitionedCheckCategoriesSpecComparisons]): Dictionary of configuration
            of checks for table comparisons. The key that identifies each comparison must match the name of a data
            comparison that is configured on the parent table.
    """

    custom: Union[Unset, "TableMonthlyPartitionedCheckCategoriesSpecCustom"] = UNSET
    volume: Union[Unset, "TableVolumeMonthlyPartitionedChecksSpec"] = UNSET
    timeliness: Union[Unset, "TableTimelinessMonthlyPartitionedChecksSpec"] = UNSET
    custom_sql: Union[Unset, "TableCustomSqlMonthlyPartitionedChecksSpec"] = UNSET
    uniqueness: Union[Unset, "TableUniquenessMonthlyPartitionChecksSpec"] = UNSET
    comparisons: Union[
        Unset, "TableMonthlyPartitionedCheckCategoriesSpecComparisons"
    ] = UNSET
    additional_properties: Dict[str, Any] = _attrs_field(init=False, factory=dict)

    def to_dict(self) -> Dict[str, Any]:
        custom: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.custom, Unset):
            custom = self.custom.to_dict()

        volume: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.volume, Unset):
            volume = self.volume.to_dict()

        timeliness: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.timeliness, Unset):
            timeliness = self.timeliness.to_dict()

        custom_sql: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.custom_sql, Unset):
            custom_sql = self.custom_sql.to_dict()

        uniqueness: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.uniqueness, Unset):
            uniqueness = self.uniqueness.to_dict()

        comparisons: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.comparisons, Unset):
            comparisons = self.comparisons.to_dict()

        field_dict: Dict[str, Any] = {}
        field_dict.update(self.additional_properties)
        field_dict.update({})
        if custom is not UNSET:
            field_dict["custom"] = custom
        if volume is not UNSET:
            field_dict["volume"] = volume
        if timeliness is not UNSET:
            field_dict["timeliness"] = timeliness
        if custom_sql is not UNSET:
            field_dict["custom_sql"] = custom_sql
        if uniqueness is not UNSET:
            field_dict["uniqueness"] = uniqueness
        if comparisons is not UNSET:
            field_dict["comparisons"] = comparisons

        return field_dict

    @classmethod
    def from_dict(cls: Type[T], src_dict: Dict[str, Any]) -> T:
        from ..models.table_custom_sql_monthly_partitioned_checks_spec import (
            TableCustomSqlMonthlyPartitionedChecksSpec,
        )
        from ..models.table_monthly_partitioned_check_categories_spec_comparisons import (
            TableMonthlyPartitionedCheckCategoriesSpecComparisons,
        )
        from ..models.table_monthly_partitioned_check_categories_spec_custom import (
            TableMonthlyPartitionedCheckCategoriesSpecCustom,
        )
        from ..models.table_timeliness_monthly_partitioned_checks_spec import (
            TableTimelinessMonthlyPartitionedChecksSpec,
        )
        from ..models.table_uniqueness_monthly_partition_checks_spec import (
            TableUniquenessMonthlyPartitionChecksSpec,
        )
        from ..models.table_volume_monthly_partitioned_checks_spec import (
            TableVolumeMonthlyPartitionedChecksSpec,
        )

        d = src_dict.copy()
        _custom = d.pop("custom", UNSET)
        custom: Union[Unset, TableMonthlyPartitionedCheckCategoriesSpecCustom]
        if isinstance(_custom, Unset):
            custom = UNSET
        else:
            custom = TableMonthlyPartitionedCheckCategoriesSpecCustom.from_dict(_custom)

        _volume = d.pop("volume", UNSET)
        volume: Union[Unset, TableVolumeMonthlyPartitionedChecksSpec]
        if isinstance(_volume, Unset):
            volume = UNSET
        else:
            volume = TableVolumeMonthlyPartitionedChecksSpec.from_dict(_volume)

        _timeliness = d.pop("timeliness", UNSET)
        timeliness: Union[Unset, TableTimelinessMonthlyPartitionedChecksSpec]
        if isinstance(_timeliness, Unset):
            timeliness = UNSET
        else:
            timeliness = TableTimelinessMonthlyPartitionedChecksSpec.from_dict(
                _timeliness
            )

        _custom_sql = d.pop("custom_sql", UNSET)
        custom_sql: Union[Unset, TableCustomSqlMonthlyPartitionedChecksSpec]
        if isinstance(_custom_sql, Unset):
            custom_sql = UNSET
        else:
            custom_sql = TableCustomSqlMonthlyPartitionedChecksSpec.from_dict(
                _custom_sql
            )

        _uniqueness = d.pop("uniqueness", UNSET)
        uniqueness: Union[Unset, TableUniquenessMonthlyPartitionChecksSpec]
        if isinstance(_uniqueness, Unset):
            uniqueness = UNSET
        else:
            uniqueness = TableUniquenessMonthlyPartitionChecksSpec.from_dict(
                _uniqueness
            )

        _comparisons = d.pop("comparisons", UNSET)
        comparisons: Union[Unset, TableMonthlyPartitionedCheckCategoriesSpecComparisons]
        if isinstance(_comparisons, Unset):
            comparisons = UNSET
        else:
            comparisons = (
                TableMonthlyPartitionedCheckCategoriesSpecComparisons.from_dict(
                    _comparisons
                )
            )

        table_monthly_partitioned_check_categories_spec = cls(
            custom=custom,
            volume=volume,
            timeliness=timeliness,
            custom_sql=custom_sql,
            uniqueness=uniqueness,
            comparisons=comparisons,
        )

        table_monthly_partitioned_check_categories_spec.additional_properties = d
        return table_monthly_partitioned_check_categories_spec

    @property
    def additional_keys(self) -> List[str]:
        return list(self.additional_properties.keys())

    def __getitem__(self, key: str) -> Any:
        return self.additional_properties[key]

    def __setitem__(self, key: str, value: Any) -> None:
        self.additional_properties[key] = value

    def __delitem__(self, key: str) -> None:
        del self.additional_properties[key]

    def __contains__(self, key: str) -> bool:
        return key in self.additional_properties
